### PR TITLE
Mvcc Concurrent Testing

### PR DESCRIPTION
**Pull request by yiyun-sj**

### CODE REVIEW

File: test/include/graph_test/graph_test.h [hunk 1]
```diff
@@ -32,6 +37,35 @@ class EmptyDBTest : public PrivateGraphTest {
     std::string getInputDir() override { KU_UNREACHABLE; }
 };
 
+class ConcurrentTest {
```
Comment (ray6080) [inline test/include/graph_test/graph_test.h:40]: `ConcurrentTest` -> `ConcurrentTestExecutor` / `ConcurrentTestRunner`.

File: test/include/graph_test/graph_test.h [hunk 2]
```diff
@@ -32,6 +37,35 @@ class EmptyDBTest : public PrivateGraphTest {
...
+public:
+    ConcurrentTest(bool& connectionsPaused, main::Connection& connection, std::string databasePath):
+      connectionPaused{connectionsPaused}, connection{connection}, databasePath{databasePath} {}
```
Comment (ray6080) [inline test/include/graph_test/graph_test.h:43]: Better leave an empty line between constructor and other functions.

File: test/include/graph_test/graph_test.h [hunk 3]
```diff
@@ -32,6 +37,35 @@ class EmptyDBTest : public PrivateGraphTest {
...
+    }
+
+private:
+    void runStatements();
+
+    bool& connectionPaused;
+    main::Connection& connection;
+    std::string databasePath;
+    std::thread connThread;
+    std::vector<TestStatement*> statementsQueue;
```
Comment (ray6080) [inline test/include/graph_test/graph_test.h:66]: I think simply `statements` is fine.

File: test/graph_test/graph_test.cpp [hunk 4]
```diff
@@ -47,5 +48,12 @@ void DBTest::createNewDB() {
     createDBAndConn();
 }
 
+void ConcurrentTest::runStatements() {
+    while (connectionPaused) {}
```
Comment (ray6080) [inline test/graph_test/graph_test.cpp:52]: hmm isn't it simpler if you only run `runStatements` when `connectionPaused` is false? so no need to busy wait here.

File: test/include/graph_test/graph_test.h [hunk 5]
```diff
@@ -32,6 +37,35 @@ class EmptyDBTest : public PrivateGraphTest {
...
 };
 
+class ConcurrentTest {
+public:
+    ConcurrentTest(bool& connectionsPaused, main::Connection& connection, std::string databasePath):
+      connectionPaused{connectionsPaused}, connection{connection}, databasePath{databasePath} {}
+    void execute() {
+        connThread = std::thread(&ConcurrentTest::runStatements, this);
+    }
+    void join() {
```
Comment (ray6080) [inline test/include/graph_test/graph_test.h:47]: Looks to me `join()` is unnecessary, I think you should always join inside `execute`, which makes the interface simpler.

File: test/include/graph_test/graph_test.h [hunk 6]
```diff
@@ -32,6 +37,35 @@ class EmptyDBTest : public PrivateGraphTest {
...
+        statementsQueue.clear();
+    }
+    void addStatement(TestStatement* statement) {
+        statementsQueue.emplace_back(statement);
+    }
+
+private:
+    void runStatements();
+
+    bool& connectionPaused;
```
Comment (ray6080) [inline test/include/graph_test/graph_test.h:62]: `connectionPaused` is also not needed if you don't busy wait inside `runStatements`.

File: test/include/graph_test/graph_test.h (hunk 1 above)
Comment (ray6080) [inline test/include/graph_test/graph_test.h:40]: Add a comment here saying that `ConcurrentTest` is not thread safe.

File: test/include/graph_test/graph_test.h [hunk 7]
```diff
@@ -66,21 +103,43 @@ class DBTest : public PrivateGraphTest {
...
     }
 
-    inline void CheckConn(std::string connName) {
-        if (connMap[connName] == nullptr) {
-            connMap[connName] = std::make_unique<main::Connection>(database.get());
-        }
-    }
+protected:
+    bool isConcurrent = false;
+    bool connectionsPaused = false;
```
Comment (ray6080) [inline test/include/graph_test/graph_test.h:140]: I wonder if you can replace these two bool fields with the enum class ConcurrentStatusFlag.

Comment (yiyun-sj) [reply test/include/graph_test/graph_test.h:140]: we probably can just keep the bool fields as they make a little more sense to me than the enum

**Review by ray6080 — APPROVED** [review]

**Pull request merged**